File: 2019/python/day12.py
import re
from itertools import combinations
from math import gcd

X_AXIS, Y_AXIS, Z_AXIS = 0, 1, 2

# moon = [x, y, z, Vx, Vy, Vz]
def get_initial_state():
    with open('../day_12.txt') as f:
        return [[int(match) for match in re.findall(r'-?\d+', line)] + [0, 0, 0] for line in f]

def simulate(moons, axes):
    for (m0, m1) in combinations(moons, 2):
        for axis in axes:
            if m0[axis] < m1[axis]:
                m0[axis+3] += 1
                m1[axis+3] -= 1
            elif m0[axis] > m1[axis]:
                m0[axis+3] -= 1
                m1[axis+3] += 1
    for m in moons:
        for axis in axes:
            m[axis] += m[axis+3]

moons = get_initial_state()
for _ in range(1000):
    simulate(moons, [X_AXIS, Y_AXIS, Z_AXIS])
total_energy = sum( sum(abs(v) for v in m[0:3]) * sum(abs(v) for v in m[3:6]) for m in moons )
print('part 1:', total_energy)


def period(axis):
    steps = 0
    moons = get_initial_state()
    state0 = [m[axis::3] for m in moons]
    while True:
        steps += 1
        simulate(moons, [axis])
        state = [m[axis::3] for m in moons]
        if state == state0:
            break
    return steps

def lcm(a, b):
    return a * b // gcd(a, b)

print('part 2:', lcm(lcm(period(X_AXIS), period(Y_AXIS)), period(Z_AXIS)))

File: 2019/go/day12.go
package main

import (
	"AdventOfCode/util"
	"fmt"
	"regexp"
	"strconv"
)

const (
	xAxis = iota
	yAxis
	zAxis
)

type vector = [3]int

type moon struct {
	position vector
	velocity vector
}

var intRegex = regexp.MustCompile(`-?\d+`)

func parse(lines []string) []moon {
	moons := make([]moon, len(lines))
	for i, l := range lines {
		for j, s := range intRegex.FindAllString(l, -1) {
			num, _ := strconv.Atoi(s)
			moons[i].position[j] = num
		}
	}
	return moons
}

func step(moons []moon) {
	for i := 0; i < len(moons)-1; i++ {
		for j := i; j < len(moons); j++ {
			for d := 0; d < 3; d++ {
				switch cmp := moons[i].position[d] - moons[j].position[d]; {
				case cmp < 0:
					moons[i].velocity[d]++
					moons[j].velocity[d]--
				case cmp > 0:
					moons[i].velocity[d]--
					moons[j].velocity[d]++
				}
			}
		}
	}

	for i := 0; i < len(moons); i++ {
		for d := 0; d < 3; d++ {
			moons[i].position[d] += moons[i].velocity[d]
		}
	}
}

func energy(moons []moon) int {
	var e int
	for i := 0; i < len(moons); i++ {
		var pe, ke int
		for d := 0; d < 3; d++ {
			pe += util.Abs(moons[i].position[d])
			ke += util.Abs(moons[i].velocity[d])
		}
		e += pe * ke
	}
	return e
}

func findPeriod(moons []moon, axis int) int {
	m := make([]moon, len(moons))
	copy(m, moons)

	p := 0
	equal := false
	for !equal {
		p++
		step(m)

		equal = true
		for i := 0; i < len(moons); i++ {
			if m[i].position[axis] != moons[i].position[axis] ||
				m[i].velocity[axis] != moons[i].velocity[axis] {
				equal = false
				break
			}
		}
	}
	return p
}

func part1(moons []moon) int {
	for i := 0; i < 1000; i++ {
		step(moons)
	}
	return energy(moons)
}

func part2(moons []moon) int {
	px := findPeriod(moons, xAxis)
	py := findPeriod(moons, yAxis)
	pz := findPeriod(moons, zAxis)
	return util.Lcm(px, util.Lcm(py, pz))
}

func main() {
	input, _ := util.ReadInput("../day_12.txt")
	moons := parse(input)

	fmt.Printf("part 1: %d\n", part1(moons))
	fmt.Printf("part 2: %d\n", part2(moons))
}
